GVRf/Framework/jni/shaders/material/unlit_fbo_shader.cpp

```diff
@@ -0,0 +1,125 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "unlit_fbo_shader.h"
+
+#include "gl/gl_program.h"
+#include "objects/material.h"
+#include "objects/mesh.h"
+#include "objects/components/render_data.h"
+#include "objects/textures/texture.h"
+#include "util/gvr_gl.h"
+
+namespace gvr {
+static const char VERTEX_SHADER[] = "attribute vec4 a_position;\n"
+        "attribute vec4 a_tex_coord;\n"
+        "uniform mat4 u_mvp;\n"
+        "varying vec2 v_tex_coord;\n"
+        "void main() {\n"
+        "  v_tex_coord.x = a_tex_coord.x;\n"
+        "  v_tex_coord.y = 1.0 - a_tex_coord.y;\n"
+        "  gl_Position = u_mvp * a_position;\n"
+        "}\n";
+
+static const char FRAGMENT_SHADER[] =
+        "precision highp float;\n"
+                "uniform sampler2D u_texture;\n"
+                "uniform vec3 u_color;\n"
+                "uniform float u_opacity;\n"
+                "varying vec2 v_tex_coord;\n"
+                "void main()\n"
+                "{\n"
+                "  vec4 color = texture2D(u_texture, v_tex_coord);"
+                "  gl_FragColor = vec4(color.r * u_color.r * u_opacity, color.g * u_color.g * u_opacity, color.b * u_color.b * u_opacity, color.a * u_opacity);\n"
+                "}\n";
+
+UnlitFboShader::UnlitFboShader() :
+        program_(0), u_mvp_(0), u_texture_(0), u_color_(0), u_opacity_(0) {
+    program_ = new GLProgram(VERTEX_SHADER, FRAGMENT_SHADER);
+    u_mvp_ = glGetUniformLocation(program_->id(), "u_mvp");
+    u_texture_ = glGetUniformLocation(program_->id(), "u_texture");
+    u_color_ = glGetUniformLocation(program_->id(), "u_color");
+    u_opacity_ = glGetUniformLocation(program_->id(), "u_opacity");
+}
+
+UnlitFboShader::~UnlitFboShader() {
+    if (program_ != 0) {
+        recycle();
+    }
+}
+
+void UnlitFboShader::recycle() {
+    delete program_;
+    program_ = 0;
+}
+
+void UnlitFboShader::render(const glm::mat4& mvp_matrix,
+        RenderData* render_data, Material* material) {
+    Mesh* mesh = render_data->mesh();
+    Texture* texture = material->getTexture("main_texture");
+    glm::vec3 color = material->getVec3("color");
+    float opacity = material->getFloat("opacity");
+
+    if (texture->getTarget() != GL_TEXTURE_2D) {
+        std::string error = "UnlitFboShader::render : texture with wrong target";
+        throw error;
+    }
+
+#if 1
+    mesh->generateVAO();
+
+    glUseProgram(program_->id());
+
+    glUniformMatrix4fv(u_mvp_, 1, GL_FALSE, glm::value_ptr(mvp_matrix));
+    glActiveTexture (GL_TEXTURE0);
+    glBindTexture(texture->getTarget(), texture->getId());
+    glUniform1i(u_texture_, 0);
+    glUniform3f(u_color_, color.r, color.g, color.b);
+    glUniform1f(u_opacity_, opacity);
+
+    glBindVertexArray(mesh->getVAOId(Material::UNLIT_FBO_SHADER));
+    glDrawElements(render_data->draw_mode(), mesh->indices().size(),
+            GL_UNSIGNED_SHORT, 0);
+    glBindVertexArray(0);
+#else
+
```

Comment:
Does this mean I can use uniform buffer objects in my shaders?

Nola

From: thomasflynn [mailto:notifications@github.com]
Sent: Tuesday, March 22, 2016 3:12 PM
To: Samsung/GearVRf <GearVRf@noreply.github.com>
Subject: Re: [GearVRf] Adding GVRF widget framework (#367)


In GVRf/Framework/jni/shaders/material/unlit_fbo_shader.cpp<https://github.com/Samsung/GearVRf/pull/367#discussion_r57078689>:

> +

> +    glUseProgram(program_->id());

> +

> +    glUniformMatrix4fv(u_mvp_, 1, GL_FALSE, glm::value_ptr(mvp_matrix));

> +    glActiveTexture (GL_TEXTURE0);

> +    glBindTexture(texture->getTarget(), texture->getId());

> +    glUniform1i(u_texture_, 0);

> +    glUniform3f(u_color_, color.r, color.g, color.b);

> +    glUniform1f(u_opacity_, opacity);

> +

> +    glBindVertexArray(mesh->getVAOId(Material::UNLIT_FBO_SHADER));

> +    glDrawElements(render_data->draw_mode(), mesh->indices().size(),

> +            GL_UNSIGNED_SHORT, 0);

> +    glBindVertexArray(0);

> +#else

> +

no need for ES2 code anymore. you can remove the #if 1 #else statements and remove the ES2 code.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub<https://github.com/Samsung/GearVRf/pull/367/files/6c4fa01a0b3d94a41c1dbf5e07b0880c8d548cf4#r57078689>